contracts/CreditVoucher.sol

```diff
@@ -0,0 +1,178 @@
+        if (paused()) {
+            if (_ownerOf(tokenId) != address(0) || to != address(0)) {
+                revert TransferWhilePaused();
+            }
+        }
+        return super._update(to, tokenId, auth);
+    }
+
+    function _baseURI() internal view override returns (string memory) {
+        return baseTokenURI;
+    }
+
+    /**
+     * @notice Set operator address
+     * @dev    Caller must be CONTRACT OWNER
+     * @param _addr The new operator address
+     */
+    function setOperator(address _addr) external onlyOwner {
+        operator = _addr;
+        emit NewOperator(_addr);
+    }
+
+    /**
+     * @notice Set base token URI
+     * @dev    Caller must be CONTRACT OWNER or OPERATOR
+     * @param _uri Th new token base URI
+     */
+    function setBaseTokenURI(string calldata _uri) external {
+        address msgSender = _msgSender();
+        if (msgSender != owner() && msgSender != operator) {
+            revert Unauthorized();
+        }
+        baseTokenURI = _uri;
+        emit NewBaseTokenURI(_uri);
+    }
+
+    /**
+        @notice Set validity duration
+        @dev Caller must be CONTRACT OWNER
+        @param _duration the new validity duration
+    */
+    function setValidityDuration(uint256 _duration) external onlyOwner {
+        validityDuration = _duration;
+        emit NewValidityDuration(_duration);
+    }
+
+    /**
+     * @notice Disable token transfer
+     * @dev Caller must be CONTRACT OWNER
+     */
+    function pause() external onlyOwner {
+        _pause();
+    }
+
+    /**
+     * @notice Enable token transfer
+     * @dev Caller must be CONTRACT OWNER
+     */
+    function unpause() external onlyOwner {
+        _unpause();
+    }
+
+    /**
+     * @notice Lazy mint a credit voucher token
+     * @dev    Caller can be ANYONE
+     * @param _to The new voucher owner
+     * @param _deadline The deadline timestamp by which the call must be mined for the approve to work
+     * @param _signature The signature provided by token owner
+     */
+    function mint(address _to, uint256 _creditValue, uint256 _deadline, bytes calldata _signature) public {
+        if (_msgSender() != _to) {
+            revert InvalidMsgSender();
+        }
+
+        bytes32 structHash = keccak256(abi.encode(MINT_TYPEHASH, _to, _creditValue, sigNonces[_to]++, _deadline));
+        bytes32 digest = _hashTypedDataV4(structHash);
+        _validateRecoveredAddress(digest, operator, _deadline, _signature);
+
+        uint256 tokenId = totalSupply() + 1;
+        _safeMint(_to, tokenId);
+        vouchers[tokenId] = Voucher(_creditValue, block.timestamp, 0, false);
+
+        emit Minted(_to, tokenId, _creditValue);
+    }
+
+    /**
+     * @notice Redeem a voucher for its pegged token equivalent
+     * @dev    Caller can be ANYONE
+     * @param _tokenId The tokenId to be redeemed
+     */
+    function redeem(uint256 _tokenId) public {
```

Comment:
The redemption logic needs to be different. check the document that I shared with you, and let's discuss in Slack.